### PR TITLE
Core: change Region caching to on_change from on-miss-strategy

### PR DESCRIPTION
**Pull request by Berserker66**
## What is this fixing or adding?
*should* make several parts of AP faster, but have not gotten around to benchmarking yet.
Difference may be negligle, possibly even worse, for small single-world gens.

## How was this tested?
unittests and some small gens. SDV is failing unittests currently and I have no yet checked why.

### CODE REVIEW

**Comment by Berserker66** [comment]
It's now failing LttP Dungeon and OoT tests, have not figured out why yet, I assume I missed something that is done to one of the variables I wrapped. This should be slower assuming all worlds used the old API perfectly, but faster in reality, though benchmarking still needs to be done with the latest changes.

**Comment by Berserker66** [comment]
Should be good for testing now. Unittests are passing, it *should* be faster and more convenient.

**Comment by Berserker66** [comment]
300 Factorio worlds:
new Region cache: 154 seconds
main: 160 seconds

Probably within noise margin. But it's also a world that already behaved in regards to the old API and this PR prevents "mistakes".

File: BaseClasses.py [hunk 1]
```diff
@@ -427,28 +425,22 @@ def push_item(self, location: Location, item: Item, collect: bool = True):
...
-            self._cached_locations = [location for region in self.regions for location in region.locations]
         if player is not None:
-            return [location for location in self._cached_locations if location.player == player]
-        return self._cached_locations
-
-    def clear_location_cache(self):
-        self._cached_locations = None
+            return list(self.regions.location_cache[player].values())
+        return list(itertools.chain(*(self.regions.location_cache[player].values()
+                                      for player in self.regions.location_cache)))
```
Comment (black-sliver) [inline BaseClasses.py:443]: This looks slow. I'm wondering if we should add a cache for this, or return an Iterator instead of a list?

The change to evermore (and other worlds) fixes the performance regression, but I'm still not convinced this is the best solution yet.

Comment (Berserker66) [reply BaseClasses.py:443]: unfortunately returning an iterator failed tests. earlier versions actually had this cached. Do we lose enough performance here that we want to use the memory to cache it?

Comment (Berserker66) [reply BaseClasses.py:443]: I guess another method is to offer iterator versions of these functions and use them where it works.

Comment (Berserker66) [reply BaseClasses.py]: Should be addressed now.

**Comment by BootsinSoots** [comment]
Test 1: 3 1-coinsanity DLCQuests because that's what I had in Players

#2366: 65.06 seconds
Main: 68.86

If I still remembered how to do t-tests, this is where I would tell you it's not significantly different. Going  test ER games next

**Comment by BootsinSoots** [comment]
Test 2: 10 default SDV template yamls

#2366: 85.2261703000404
Main: 84.28835940011777

Testing with ER turned on next (but not chaos because no)

**Comment by black-sliver** [comment]
You gotta roll with set seed (`python Generate.py --seed 123456`) and ideally turn off spoiler, because that's not fully deterministic and also the variance in cpu clock can skew the result.

**Comment by CodeTriangle** [comment]
I'm the beta world developer for CrossCode. In a test with 1000 random worlds. When run on main, time spent in create_regions would creep up by a factor of about 160 by the end of the generation. With #2366 merged into the local branch, time spent in the function shot down to an average of 0.0027 seconds as opposed to the 0.1376 seconds average when generated on main (these were tested with the same seed).

In short, this makes all the difference for my world.

**Comment by BootsinSoots** [comment]
> You gotta roll with set seed (`python Generate.py --seed 123456`) and ideally turn off spoiler, because that's not fully deterministic and also the variance in cpu clock can skew the result.

Ah, I thought the point was to test it without determinism. I can change my settings then

**Comment by BootsinSoots** [comment]
Test 3: 10 SDV Full ER seeds, seed # 123456

main: 115.04147580009885 seconds
2336: 116.55106319999322 seconds

Difference appears to be minimal, as predicted, in small worlds. Will test with 100 worlds next

**Comment by kindasneaki** [comment]
Tests for RoR2: seed # 70986686017450515584

100 world:
main: 27.220027624978684
2366: 25.535646124975756

300 world:
main: 170.74974070803728
2366: 156.1382752499776

10 world.
main: 0.7691042909864336
2366: 0.7513600420206785

**Comment by black-sliver** [comment]
ap-roller with 3 yamls per run and `-O`, `--spoiler 0` shows an improvement of ~5% after 2500 runs for `soe,v6,spire,subnautica,zillion,timespinner,smz3,sm64ex,sc2wol,sa2b,ror2,rogues_legacy,raft,minecraft,meritous,hk,factorio,dark_souls_3,archipidle,alttp`

**Comment by black-sliver** [comment]
After letting it run for a bit longer, the "up to 12 yamls" ended up not being faster on average. I guess that means it very much depends on the game, and most of the games above are not affected.

**Comment by black-sliver** [comment]
At this point, I'm wondering what cythonizing the containers would do

**Comment by Berserker66** [comment]
> At this point, I'm wondering what cythonizing the containers would do

might be worth a shot

**Review by black-sliver — APPROVED** [review]
Looks like it's not slower on average.
